### DISCUSSION TITLE
ImageSharp

### DISCUSSION THREAD

**Discussion by elijahr2411** [MarcusWichelmann/MarcusW.VncClient#17] [Q&A] [answered]
Hi! Do you have any pointers on getting your library to work with ImageSharp? I've read through the documentation and am honestly quite confused, it'd be great if you could help a bit. My goal is to, on each frame update, render the frame into an ImageSharp image for use with my program. Thanks!

**Comment by MarcusWichelmann** [comment, marked as answer]
Hi, interesting idea.
You could do that by implementing your own `IRenderTarget`. For that, please see the [Readme](https://github.com/MarcusWichelmann/MarcusW.VncClient#manual-integration) and this Avalonia-specific sample implementation:
https://github.com/MarcusWichelmann/MarcusW.VncClient/blob/master/src/MarcusW.VncClient.Avalonia/RfbRenderTarget.cs#L31

You would only have to implement the `GrabFramebufferReference` method. As you can see, there it creates a `WriteableBitmap`, which is also Avalonia-specific, but could be any other Bitmap implementation. All the VNC-Client implementation needs, is a pointer to the memory, where *whatever* Bitmap implementation stores its image.

So after looking at the [ImageSharp documentation](https://docs.sixlabors.com/articles/imagesharp/pixelbuffers.html), your best bet would probably be, to allocate your own `byte[] rgbaBytes` with the required size for all the pixels, return a `IFramebufferReference` with a pointer to that byte-array and then use the Dispose method of your `IFramebufferReference? implementation (see [here](https://github.com/MarcusWichelmann/MarcusW.VncClient/blob/9d9fbf7b97b390ca9f0dff68ba52d2b9a85aed09/src/MarcusW.VncClient.Avalonia/Adapters/Rendering/AvaloniaFramebufferReference.cs)) to pass the byte-array to ImageSharp using a callback when the rendering is finished:
```c#
using (var image = Image.LoadPixelData<Rgba32>(rgbaBytes, width, height))
{
    // Work with the image
}
```

Hope that helps.

**Reply by elijahr2411** [reply]
Hi! Do you know how exactly I would pass a pointer to my byte array? The following code throws a `Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.`:

```
    public unsafe sealed class VNCFramebufferReference : IFramebufferReference {
        byte[] rgbaArray;
        Memory<byte> rgbaArrMem => rgbaArray.AsMemory();
        MemoryHandle pinHandler => rgbaArrMem.Pin();
        Size size;
        public IntPtr Address => (IntPtr)pinHandler.Pointer;
        public Size Size => size;
        public PixelFormat Format => new PixelFormat("Plain RGBA", 32, 32, bigEndian: false, trueColor: true, hasAlpha: true, 255, 255, 255, 255, 24, 16, 8, 0);
        public double HorizontalDpi => Size.Height;
        public double VerticalDpi => Size.Width;
        VNCController parentController;
        internal VNCFramebufferReference(byte[] rgbaArray, Size size, VNCController parentController) {
            this.rgbaArray = rgbaArray;
            this.size = size;
            this.parentController = parentController;
        }
        public void Dispose() {
            pinHandler.Dispose();
            parentController.renderFrame(rgbaArray, size);
        }
    }
```

**Reply by MarcusWichelmann** [reply]
I'd have went with https://docs.microsoft.com/dotnet/csharp/language-reference/keywords/fixed-statement, but Memory<T>.Pin should also work, but I haven't used that yet and am not sure if its usage is completely correct.

**Reply by elijahr2411** [reply]
I'm a bit confused on how to implement that since it seems fixed pointers are only valid from within the fixed block

I also just tried using Marshal.Copy and got the same exception

**Reply by elijahr2411** [reply]
Figured it out, the problem was a bit stupid. I was only allocating enough bytes for 1 byte-per-pixel when I needed 4, causing the memory access violation.

**Reply by MarcusWichelmann** [reply]
> I'm a bit confused on how to implement that since it seems fixed pointers are only valid from within the fixed block
Oh, you're right.

Great, that you found a fix.

**Reply by elijahr2411** [reply]
Hey, I've been working on this and have one more question. Do you know how I would detect the pixel format of the framebuffer? Because currently using `PixelFormat.Plain` gives this output.
![image](https://user-images.githubusercontent.com/62162399/181389516-5852458a-60ce-4d22-916a-d2dad36ed756.jpg)

This is what it should look like, for reference
![image](https://user-images.githubusercontent.com/62162399/181390238-d2ad1d7e-8c11-4234-99d6-a249bca9df70.png)

**Reply by MarcusWichelmann** [reply]
Then you should create a new pixel format that matches, how ImageSharp expects the pixels to exist in memory. This should be visible from the ImageSharp documentation, or you just experiment a bit. Remember to think about little/big endiannes.
